src/algorithms/all_paths.c

```diff
@@ -0,0 +1,90 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Apache License, Version 2.0,
+* modified with the Commons Clause restriction.
+*/
+
+#include <assert.h>
+#include "./all_paths.h"
+#include "../util/arr.h"
+
+void _AllPaths
+(
+    const Graph *g,         // Graph traversed.
+    Node *frontier,         // Last node reached on path.
+    int relationID,         // Edge type to traverse.
+    GrB_Matrix relation,    // Edge matrix.
+    unsigned int minHops,   // Path minimum length.
+    float maxHops,          // Path max length.
+    float hop,              // Path current length.
+    GrB_Matrix visited,     // Edges visited on path.
+    Path *path,             // Current path.    
+    Path **paths            // Paths constructed.
+)
+{
+        if(hop >= minHops && hop <= maxHops) {
+            Path clone = Path_clone(*path);
+            array_append(*paths, clone);
+        }
+
+        if(hop >= maxHops) {
+            // We've over extended the path.
+            return;
+        }
+
+        NodeID frontierID = frontier->id;
+        Vector *outGoingEdges = NewVector(Edge*, 32);
+        Graph_GetNodeEdges(g, frontier, outGoingEdges, GRAPH_EDGE_DIR_OUTGOING, relationID);
+        size_t outGoingEdgesCount = Vector_Size(outGoingEdges);
+
+        for(size_t i = 0; i < outGoingEdgesCount; i++) {
+            Edge *e;
+            Vector_Get(outGoingEdges, i, &e);
+            NodeID neighborID = Edge_GetDestNodeID(e);
+            
+            // See if edge frontier->neighbor been used.
+            bool used = false;
+            GrB_Matrix_extractElement_BOOL(&used, visited, neighborID, frontierID);
+            if(used) continue;
+
+            // Mark edge as visited.
+            GrB_Matrix_setElement_BOOL(visited, true, neighborID, frontierID);
+
+            *path = Path_append(*path, e);
+
+            _AllPaths(g, Edge_GetDestNode(e), relationID, relation, minHops, maxHops, hop+1, visited, path, paths);
+
+            Path_pop(*path);
+
+            // Mark edge as unvisited.
+            GrB_Matrix_setElement_BOOL(visited, false, neighborID, frontierID);
```

Comment:
This probably necessitates a zero vector or `GxB_NONZERO` reduction later on, right?

Reply:
Actually, never mind - I suppose it does not so long as we never try to iterate over visited with a TuplesIter or similar.

Reply:
visited, won't be iterated, it only serves as a marker for checking if an edge been traversed.